### DISCUSSION TITLE
Send SIGWINCH on theme change

### DISCUSSION THREAD

**Discussion by zoriya** [ghostty-org/ghostty#2755] [Feature Requests, Ideas] [answered]
This feature was initially created by [iterm](https://gitlab.com/gnachman/iterm2/-/issues/9855) to allow apps to update their theme when the terminal's theme changes.

This feature was then adopted by [tmux](https://github.com/tmux/tmux/issues/3582) to clear their OSC11 cache. Without this to invalidate their OSC11 cache, they will report the initial theme of the terminal and not the updated one.

**Comment by mitchellh** [comment]
cc @gpanders @rockorager 

This seems reasonable. I converted to a discussion to gather more information because I’m not totally sure yet.

one thing I’m curious about: is this sent on any palette or color change? Terminals don’t really have a notion of a theme.

or are you talking about system them? In that case I don’t see it hurting but we already support more 2031 (I think that’s what it is) for system theme change notification. It doesn’t hurt to also send a signal.

**Reply by zoriya** [reply]
I think it would make sense to send it every time the background-color or text-color is changed, since that's what the OSC11 & OSC10 tracks/reports.

I don't know what more 2031 is tho, do you have a ref?

**Reply by mitchellh** [reply]
Mode 2031: https://github.com/ghostty-org/ghostty/pull/1444

**Reply by gpanders** [reply]
One comment from the tmux discussion:

>Edit: On second thought, SIGWINCH may be triggered multiple times per second on a real window resize. Repeatedly polling for fg/bg colors could cause performance problems.

I think this is a good point. All else equal, having the ecosystem standardize around Mode 2031 instead of SIGWINCH seems better (for similar reasons that @rockorager outlined for in-band resize events). My only concern with Ghostty supporting this is that it sends a signal to app developers that SIGWINCH is the "right" way to handle this ("iTerm2, tmux, and now Ghostty all do it, so why not?").

**Reply by zoriya** [reply]
That's fair. On the other hand, it means osc-10/osc-11 are currently the most used way to detect terminal color theme (used by vim/nvim/bat/delta, idk about others). Since mode 2031 reflects system theme and not the terminal background, it could be inaccurate.

Some apps might want to know about background-color changes, even if the system theme has not changed (even tho ghostty doesn't currently support this feature).

**Reply by gpanders** [reply]
>Some apps might want to know about background-color changes

Doesn't Mode 2031 support this? From the spec:

>In order to not make the terminal emulator look bad after such switch, we must enable the applications inside the terminal to detect when the terminal has updated the color palette. This may happen either due to the operating system having changed the current theme or simply because the user has explicitly requested to reconfigure the currently used theme.

My understanding of that is that the terminal emulator should be sending the color change notifications when the terminal's color palette changes:

>Send CSI ? 2031 h to the terminal to enable unsolicited DSR (device status report) messages for color palette updates and CSI ? 2031 l respectively to disable it again.

**Reply by gpanders** [reply]
So if I'm an application author that wants to support color change detection, I can just set up an event handler for any time I receive `CSI ? 997 ; 1 n` or `CSI ? 997 ; 2 n`. If all I care about is "light or dark" then I'm done, but if I want to know the RGB colors I can then use OSC 10/11.

**Reply by mitchellh** [reply]
> Doesn't Mode 2031 support this? From the spec:

We may not have implemented this part of the spec!

**Reply by rockorager** [reply]
> My only concern with Ghostty supporting this is that it sends a signal to app developers that SIGWINCH is the "right" way to handle this ("iTerm2, tmux, and now Ghostty all do it, so why not?").

I have this same concern. This risks becoming a defacto "standard" way of reporting theme/color changes.

**Reply by mitchellh** [reply, marked as answer]
> We may not have implemented this part of the spec!

See #2771

**Reply by will** [reply]
> I have this same concern. This risks becoming a defacto "standard" way of reporting theme/color changes.

As the person who asked iterm to add that feature: sorry! I didn't know any better at the time
